rust/src/trans/sum/int/split/mod.rs

```diff
@@ -0,0 +1,114 @@
+use std::cmp::Ordering;
+
+use crate::{
+    error::Fallible, 
+    core::{Transformation, Function, StabilityRelation}, 
+    dom::{VectorDomain, BoundedDomain, AllDomain, SizedDomain},
+    dist::{SymmetricDistance, AbsoluteDistance, IntDistance}, 
+    traits::{DistanceConstant, CheckNull, AlertingAbs, InfCast, SaturatingAdd, InfSub}
+};
+
+use super::AddIsExact;
+
+#[cfg(feature = "ffi")]
+mod ffi;
+
+pub trait SplitSatSum: Sized {
+    /// Method which takes an iterator and generates `Self` from the elements by
+    /// "summing up" the items.
+    fn split_sat_sum(v: &Vec<Self>) -> Self;
+}
+
+macro_rules! impl_unsigned_int_split_sat_sum {
+    ($($ty:ty)+) => ($(impl SplitSatSum for $ty {
+        fn split_sat_sum(v: &Vec<Self>) -> Self {
+            v.iter().fold(0, |sum, v| sum.saturating_add(*v))
+        }
+    })+);
+}
+macro_rules! impl___signed_int_split_sat_sum {
+    ($($ty:ty)+) => ($(impl SplitSatSum for $ty {
+        fn split_sat_sum(v: &Vec<Self>) -> Self {
+            let (neg, pos) = v.iter().fold((0, 0), |(neg, pos), v| {
+                match v.cmp(&0) {
+                    Ordering::Less => (neg.saturating_add(&v), pos),
+                    Ordering::Greater => (neg, pos.saturating_add(&v)),
+                    Ordering::Equal => (neg, pos),
+                }
+            });
+            neg.saturating_add(pos)
+        }
+    })+);
+}
+
+impl_unsigned_int_split_sat_sum! { u8 u16 u32 u64 u128 usize }
+impl___signed_int_split_sat_sum! { i8 i16 i32 i64 i128 isize }
```

Comment:
Similar to our discussion about the implementation of `MakeBoundedSum` and `MakeSizedBoundedSum` traits, I find it a bit annoying that we have to resort to macros here, as it makes it harder to follow the flow. I understand why, and don't see any cleaner way to do it. It's just frustrating that Rust doesn't have a way to implement a trait for different generic types in different ways.

Reply:
Agreed. They don't even permit two blanket impls for generics that are provably disjoint by the type system via associated types.

```rust
struct Floats;
struct Ints;
trait Distingisher {type Assoc;}
impl Distinguisher for f64 {type Assoc = Floats;}
impl Distinguisher for i64 {type Assoc = Ints;}

impl<T: MyTrait<Assoc=Floats>> MakeBoundedSum for T {...}
impl<T: MyTrait<Assoc=Ints>> MakeBoundedSum for T {...}
```